crates/matrix-sdk-crypto/src/error.rs

```diff
@@ -70,8 +72,26 @@ pub enum OlmError {
             have a valid Olm session with us"
     )]
     MissingSession,
+
+    #[error(transparent)]
+    /// The room key that was to be shared was not shared because the sharing
+    /// strategy could not be fulfilled.
+    RoomKeySharingStrategyError(DeviceCollectError),
 }
 
+/// Depending on the sharing strategy for room keys, the distribution of the
+/// room key could fail.
+#[derive(Error, Debug)]
+pub enum DeviceCollectError {
```

Comment:
I think it would make more sense to rename this, and the enum entry in `OlmError`, to `SessionRecipientCollectionError`, to make it represent the sorts of errors that can happen when finding recipients.

Ultimately we could have `collect_session_recipients` return a `SessionRecipientCollectionError` rather than a generic `OlmError`, but that is a bigger job that should happen another day.